src/java/org/apache/cassandra/config/JMXServerOptions.java

```diff
@@ -0,0 +1,240 @@
+                                    from.jmx_encryption_options, from.login_config_name, from.login_config_file, from.password_file,
+                                    from.access_file, from.authorizer);
+    }
+
+    public JMXServerOptions(Boolean enabled,
+                            Boolean remote,
+                            int jmxPort,
+                            int rmiPort,
+                            Boolean authenticate,
+                            EncryptionOptions jmx_encryption_options,
+                            String loginConfigName,
+                            String loginConfigFile,
+                            String passwordFile,
+                            String accessFile,
+                            String authorizer)
+    {
+        this.enabled = enabled;
+        this.remote = remote;
+        this.jmx_port = jmxPort;
+        this.rmi_port = rmiPort;
+        this.authenticate = authenticate;
+        this.jmx_encryption_options = jmx_encryption_options;
+        this.login_config_name = loginConfigName;
+        this.login_config_file = loginConfigFile;
+        this.password_file = passwordFile;
+        this.access_file = accessFile;
+        this.authorizer = authorizer;
+    }
+
+    @Override
+    public String toString()
+    {
+        // we are not including encryption options on purpose
+        // as that contains credentials etc.
+        String jmxOptionsString;
+        if (jmx_encryption_options == null)
+            jmxOptionsString = "unspecified";
+        else
+            jmxOptionsString = jmx_encryption_options.enabled ? "enabled" : "disabled";
+
+        return "JMXServerOptions{" +
+               "enabled=" + enabled +
+               ", remote=" + remote +
+               ", jmx_port=" + jmx_port +
+               ", rmi_port=" + rmi_port +
+               ", authenticate=" + authenticate +
+               ", jmx_encryption_options=" + jmx_encryption_options +
+               ", login_config_name='" + login_config_name + '\'' +
+               ", login_config_file='" + login_config_file + '\'' +
+               ", password_file='" + password_file + '\'' +
+               ", access_file='" + access_file + '\'' +
+               ", authorizer='" + authorizer + '\'' +
+               '}';
+    }
+
+    public static boolean isEnabledBySystemProperties()
+    {
+        return CASSANDRA_JMX_REMOTE_PORT.isPresent() || CASSANDRA_JMX_LOCAL_PORT.isPresent();
+    }
+
+    public static JMXServerOptions createParsingSystemProperties()
+    {
+        int jmxPort;
+        boolean remote;
+        if (CASSANDRA_JMX_REMOTE_PORT.isPresent())
+        {
+            jmxPort = CASSANDRA_JMX_REMOTE_PORT.getInt();
+            remote = true;
+        }
+        else
+        {
+            jmxPort = CASSANDRA_JMX_LOCAL_PORT.getInt(7199);
+            remote = false;
+        }
+
+        boolean enabled = isEnabledBySystemProperties();
```

Comment:
I think it might be a little cleaner this way (to avoid the duplicated call of CASSANDRA_JMX_REMOTE_PORT.isPresent() and CASSANDRA_JMX_LOCAL_PORT.isPresent())

`public static boolean isEnabledBySystemProperties()
    {
        return getSystemPropertyPortAndMode() != null;
    }

    public static Pair<Integer, Boolean> getSystemPropertyPortAndMode()
    {
        int jmxPort = CASSANDRA_JMX_LOCAL_PORT.getInt(7199);
        boolean remote = false;

        if (CASSANDRA_JMX_REMOTE_PORT.isPresent()) {
            jmxPort = CASSANDRA_JMX_REMOTE_PORT.getInt();
            remote = true;
        }

        return (CASSANDRA_JMX_LOCAL_PORT.isPresent() || remote) ? Pair.create(jmxPort, remote) : null;
    }

    public static JMXServerOptions createParsingSystemProperties()
    {
        boolean enabled = false;
        boolean remote = false;
        int jmxPort = 7199;

        Pair<Integer, Boolean> portAndMode = getSystemPropertyPortAndMode();
        if (portAndMode != null) {
            enabled = true;
            jmxPort = portAndMode.left;
            remote = portAndMode.right;
        }`

Reply:
Matter of preference, I think the current way is just fine.

Reply:
I personally find Cheng's approach a bit cleaner (less duplication) but given the context no objection to leaving it as is